### PR TITLE
chore: integrate currency input widget

### LINKED ISSUE #27747: Add functionality to WDS TextInput to support CurrencyInput widget
When adding functionality, follow the [audit](https://www.notion.so/appsmith/Widgets-Properties-Audit-3832f6a7be5f47cc9bfa3bd1ddae31fa)

### PR DESCRIPTION
**Pull request by jsartisan**
Fixes #27747 

https://github.com/appsmithorg/appsmith/assets/6636360/9234acc1-f41a-4ca8-aaf6-15c66e2b3e9d

### CODE REVIEW

File: app/client/packages/design-system/headless/src/components/Field/src/Label.tsx [hunk 1]
```diff
@@ -6,37 +6,69 @@ import type { SpectrumLabelProps } from "@react-types/label";
 
 export type LabelProps = Omit<
   SpectrumLabelProps,
-  | "necessityIndicator"
```
Comment (jsartisan) [inline app/client/packages/design-system/headless/src/components/Field/src/Label.tsx:9]: Note: this is part of the refactoring I had to do to make the component little simpler. We were previously defining the necessityIndicator and a few other properties in the styled version of the component. It was creating confusion why few properties are in headless and few in styled. Main reason for it was that we couldn't styled text using className but now we can use `getTypographyClassName` to style any text using className.
Now we are moving these properties back to headless and we are styling description and label with LabelClassName and HelptextClassName

File: app/client/packages/design-system/widgets/src/components/TextArea/src/TextArea.tsx [hunk 2]
```diff
@@ -6,71 +6,42 @@ import type {
...
-      label={label}
-      necessityIndicator={necessityIndicator}
-    />
-  );
-
   const contextualHelp = Boolean(contextualHelpProp) && (
     <ContextualHelp contextualHelp={contextualHelpProp} />
   );
 
-  const wrappedDescription = Boolean(description) && (
```
Comment (jsartisan) [inline app/client/packages/design-system/widgets/src/components/TextArea/src/TextArea.tsx:54]: This was the main reason of separating few props in styled version and few in this headless. previously we have to use the text component to style the label and description, so we needed to wrap them. But now we can style them using classNames. so this is not needed anymore.

**Review by KelvinOm — CHANGES_REQUESTED** [review]
Decimals allowed — seems like it doesn't work.
Required — doesn't add a mandatory attribute `*`

Do not forget to enable `release_anvil_enabled ` flag for testing.

**Review by dhruvikn — CHANGES_REQUESTED** [review]
Found some issues while testing the widget. 

1. The currency icon is getting overflowed. Select AED and resize to reproduce.

<video src="https://github.com/appsmithorg/appsmith/assets/22471214/17643055-f2e8-4e85-a336-33bd5f2972a8"></video>

2. Allow currency change property doesn't do anything in both edit and deploy mode. Tried changing the currency by clicking the icon but nothing happened. Am I missing anything here?

File: app/client/packages/design-system/widgets/src/styles/src/text-input.module.css [hunk 3]
```diff
@@ -5,7 +5,7 @@
     align-items: center;
     gap: var(--inner-spacing-1);
     border-radius: var(--border-radius-1);
-    box-shadow: 0 0 0 1px var(--color-bd-neutral);
+    box-shadow: inset 0 0 0 1px var(--color-bd-neutral);
```
[the remark this answers is not in the document]
Comment (jsartisan) [reply app/client/packages/design-system/widgets/src/styles/src/text-input.module.css:8]: The input was looking bigger when placed next to a button. so used the inset box shadow.

File: app/client/packages/design-system/widgets/src/components/TextInput/stories/TextInput.stories.mdx [hunk 4]
```diff
@@ -67,7 +67,7 @@ TextInput is a component that allows users to input text.
   name="Prefix"
   args={{
     placeholder: "Start Visual",
-    startIcon: "$",
+    startIcon: $,
```
[the remark this answers is not in the document]
Comment (jsartisan) [reply app/client/packages/design-system/widgets/src/components/TextInput/stories/TextInput.stories.mdx:70]: thanks for catching this. it was not wokring. fixed it.

File: app/client/src/widgets/wds/WDSCurrencyInputWidget/widget/index.tsx [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+    );
+  }
+
+  static getPropertyPaneStyleConfig() {
+    return super.getPropertyPaneStyleConfig();
+  }
+
+  static getDerivedPropertiesMap() {
+    return {
+      isValid: `{{(()=>{${derivedProperties.isValid}})()}}`,
```
[the remark this answers is not in the document]
Comment (jsartisan) [reply app/client/src/widgets/wds/WDSCurrencyInputWidget/widget/index.tsx:84]: This is how it is designed.

File: app/client/src/widgets/wds/WDSCurrencyInputWidget/component/types.ts [hunk 6]
```diff
@@ -0,0 +1,9 @@
+import type { BaseInputComponentProps } from "widgets/wds/WDSBaseInputWidget";
+
+export interface CurrencyInputComponentProps extends BaseInputComponentProps {
+  currencyCode?: string;
+  noOfDecimals?: number;
+  allowCurrencyChange?: boolean;
+  decimals?: number;
+  onCurrencyChange: (code?: any) => void;
```
[the remark this answers is not in the document]
Comment (jsartisan) [reply app/client/src/widgets/wds/WDSCurrencyInputWidget/component/types.ts:8]: ✅

**Review by KelvinOm — DISMISSED** [review]
Code is LGTM. One question about UX.

Previously, we didn't allow to enter anything other than numbers and comma, now we can enter anything, but the value is reset obBlur event. Is this expected/acceptable behavior?


https://github.com/appsmithorg/appsmith/assets/11555074/6f46ef98-ca21-43ea-bbed-2e797c0b88be

cc @ichik

**Comment by ichik** [comment]
It's better to show an error than to erase user's input.

**Comment by ichik** [comment]
Why do we list Euro for each individual country that uses it?
Also, please be aware that not every convention uses comma, some use point: https://en.wikipedia.org/wiki/Decimal_separator#Hindu–Arabic_numerals

**Comment by dhruvikn** [comment]
> Why do we list Euro for each individual country that uses it? Also, please be aware that not every convention uses comma, some use point: [https://en.wikipedia.org/wiki/Decimal_separator#Hindu–Arabic_numerals](https://en.wikipedia.org/wiki/Decimal_separator#Hindu%E2%80%93Arabic_numerals)

If we are solving for this, I think it's also worth mentioning that the place values used in the Indian numeral system are different than the International numeral system. 

Example - 1000000 would be 10,00,000 in the Indian system and 1,000,000 in the international one.

image

**Comment by ichik** [comment]
Yes, there's a huge variety with digit grouping too, but at least that is done automatically and doesn't interfere with user's input / result in an error unlike the decimal.

**Comment by jsartisan** [comment]
> Why do we list Euro for each individual country that uses it?
I am also not sure about why did this way.

> Also, please be aware that not every convention uses comma, some use point:
Yes, that is handled with Intl class ( https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl )

**Comment by jsartisan** [comment]
>  If we are solving for this, I think it's also worth mentioning that the place values used in the Indian numeral system are different than the International numeral system.

Very good catch. The current implementation is not solving this. But let me create an issue for this to track.

**Comment by jsartisan** [comment]
@dhruvikn added the issue here - https://github.com/appsmithorg/appsmith/issues/28862

**Comment by ichik** [comment]
> Yes, that is handled with Intl class ( https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl )

That seems unreliable, could we instead allow both comma and period as valid input?

**Comment by jsartisan** [comment]
> That seems unreliable, could we instead allow both comma and period as valid input?
It should depend on the language the user is using. Allowing both will just create confusion.

**Comment by ichik** [comment]
> It should depend on the language the user is using. Allowing both will just create confusion.

That is very unreliable, subject to a combination of many OS and browser settings. It is very likely therefore to not match what the user is expecting. Allowing both would only create confusion if user is trying to do digit grouping manually, but at least would allow decimal input.

**Comment by jsartisan** [comment]
I feel you are mixing two things, Language and Currency. How we show the currency should depend on the language user is using. 🤔
Currency is just an number but the period and comma representation are the part of the language.

**Review by KelvinOm — APPROVED** [review]

**Pull request merged**